File: 12. BST 4.py
class Node:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None

class BuildTree:
    def __init__(self):
        self.idx = 0

    def buildTree(self, arr):
        if self.idx >= len(arr):
            return None
        if arr[self.idx] == -1:
            self.idx += 1
            return None
        root = Node(arr[self.idx])
        self.idx += 1
        root.left = self.buildTree(arr)
        root.right = self.buildTree(arr)
        return root

    def maxDepth(self, node):
        if node is None:
            return 0
        left_depth = self.maxDepth(node.left)
        right_depth = self.maxDepth(node.right)

        return 1 + max(left_depth, right_depth)


arr = list(map(int, input().split()))
bt = BuildTree()
root = bt.buildTree(arr)
print(bt.maxDepth(root))

File: 12. BST 4.java
import java.util.*;

class Node {
    int data;
    Node left;
    Node right;

    Node(int data) {
        this.data = data;
        this.left = null;
        this.right = null;
    }
}

class BuildTree {
    int idx = 0;

    public Node buildTree(int[] arr) {
        if (idx >= arr.length || arr[idx] == -1) {
            idx++;
            return null;
        }

        Node root = new Node(arr[idx++]);
        root.left = buildTree(arr);
        root.right = buildTree(arr);

        return root;
    }
    public int maxDepth(Node root) {
        if (root == null)
            return 0;

        int leftDepth = maxDepth(root.left);
        int rightDepth = maxDepth(root.right);

        return Math.max(leftDepth, rightDepth) + 1;
    }
}



public class Main {
    public static void main(String[] args) {

        Scanner sc = new Scanner(System.in);
        String line = sc.nextLine();
        String[] tokens = line.split(" ");

        int[] arr = new int[tokens.length];
        for (int i = 0; i < tokens.length; i++) {
            arr[i] = Integer.valueOf(tokens[i]);
        }

        BuildTree bt = new BuildTree();
        Node root = bt.buildTree(arr);

        System.out.println(bt.maxDepth(root));
    }
}
